LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs

```diff
@@ -65,7 +65,7 @@ private string Message(DownlinkMessage message)
             writer.WriteNumber("dC", (int)deviceClassType);
 
             // Getting and writing payload bytes
-            var pduBytes = message.Data;
+            var pduBytes = message.Data.ToArray();
```

Comment:
I think you can simply use the Data directly?

Reply:
you would have to add an overload to Hexadecimal to deal with the memory or use the Span (exposed from ReadOnlyMemory<>).

In Hexadecimal for example:

`public static void Write(ReadOnlyMemory<byte> buffer, Span<char> output) => Write(buffer.Span, output, null);`

Reply:
Good point!

Reply:
I think what @p-schuler meant (as a better alternative) is that you can just use `Span` on the memory and use the existing methods of `Hexadecimal` instead of needing to add an overload for `Write`:

```diff
diff --git a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs
index 6ce70588..d5b409c1 100644
--- a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs
+++ b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs
@@ -67,7 +67,7 @@ namespace LoRaWan.NetworkServer.BasicsStation
             // Getting and writing payload bytes
             var pduBytes = message.Data;
             var pduChars = new char[pduBytes.Length * 2];
-            Hexadecimal.Write(pduBytes, pduChars);
+            Hexadecimal.Write(pduBytes.Span, pduChars);
             writer.WriteString("pdu", pduChars);
 
 #pragma warning disable CA5394 // Do not use insecure randomness. This is fine as not used for any crypto operations.
diff --git a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan/Hexadecimal.cs b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan/Hexadecimal.cs
index 0498e4a1..77b6abb8 100644
--- a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan/Hexadecimal.cs
+++ b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan/Hexadecimal.cs
@@ -19,8 +19,6 @@ namespace LoRaWan
                 throw new ArgumentException(InsufficientBufferSizeErrorMessage, paramName);
         }
 
-        public static void Write(ReadOnlyMemory<byte> buffer, Span<char> output) => Write(buffer.Span, output, null);
-
         public static Span<char> Write(byte value, Span<char> output, LetterCase letterCase = LetterCase.Upper)
         {
             ValidateSufficientlySizedBuffer(output.Length, sizeof(byte), nameof(output));
```

I wouldn't favour adding an overload because we'd have to technically maintain more code and unit tests for those overloads (missing in this PR).